core/queryalgebra/model/src/main/java/org/eclipse/rdf4j/query/algebra/QueryModelNode.java

```diff
@@ -73,16 +75,33 @@
 	/**
 	 * Returns the signature of this query model node. Signatures normally include the node's name and any parameters,
 	 * but not parent or child nodes. This method is used by {@link #toString()}.
-	 * 
+	 *
 	 * @return The node's signature, e.g. <tt>SLICE (offset=10, limit=10)</tt>.
 	 */
 	public String getSignature();
 
 	/**
 	 * Returns a (deep) clone of this query model node. This method recursively clones the entire node tree, starting
 	 * from this nodes.
-	 * 
+	 *
 	 * @return A deep clone of this query model node.
 	 */
 	public QueryModelNode clone();
+
+	/**
+	 * Returns the number of tuples that this QueryNode predicts will be outputted. For a StatementPattern this would be
+	 * the estimated cardinality provided by the EvaluationStatistics. For a Join the would be the resulting number of
+	 * joined tuples.
+	 *
+	 * @return rows
+	 */
+	@Experimental
+	default double getEstimatedRows() {
+		return -1;
+	}
+
+	@Experimental
+	default void setEstimatedRows(double rows) {
+		// no-op for backwards compatibility
+	}
```

Comment:
Propose we use the PostgreSQL naming of "rows" for this cardinality. I've prefixed it with "estimated" to make it more obvious what we are talking about.

In order for this to be backwards compatible I propose that setEstimatedRows be a no-op. We can remove both default implementations in 4.0.0, I'm just a bit unsure now how we can mark these methods to make it obvious for anyone who stumbles upon them. I have marked them Experimental though, since we might want to change the name or something else in the future. I'm open for other names though, "getEstimatedTuples" comes to mind.

Reply:
Couple of things:

1. Rows is SQL/table-speak. SPARQL doesn't _have_ rows or tables (certainly construct queries don't). It has solutions, and solution sequences. 

2. You're interpreting cardinality/cost to always be the number of solutions. That's just not true. 

I'd stick with the more abstract 'costEstimate' for this. This gives us freedom to use any  relevant measure (including number of solutions) as a cost estimate.

Reply:
Happy with the noop solution though, that makes sense.

Reply:
Rows could also be interpreted as rows of solutions. I think “tuples” is better. 

Cost is different from tuples. Cost is a calculated field based on a cost model. It could take into account the difference between a request to disk vs network or a sort+mergejoin vs hashjoin. 

Tuples is an estimated ceiling. It doesn’t account for what is likely to happen, for instance based on bound variables or a limit higher up in the algebra model. A join operation should account for bound variables. Consider the tuples estimate to be what this single operation would produce if only that node with its sub tree existed.

Reply:
If we want to introduce the abstract concept of cost, we should add a new field for that.

Reply:
> Rows could also be interpreted as rows of solutions. I think “tuples” is better.
> 
> Cost is different from tuples. Cost is a calculated field based on a cost model. It could take into account the difference between a request to disk vs network or a sort+mergejoin vs hashjoin.

...which is precisely what the (misnamed) cardinality calculator currently does. For example it assigns a higher cardinality score to a SERVICE clause, not because there are more solutions expected, but because it tries to account for network latency. So what the cardinality calculator gives you back is an estimated cost, _not_ an estimated number of solutions.

> Tuples is an estimated ceiling. It doesn’t account for what is likely to happen, for instance based on bound variables or a limit higher up in the algebra model. A join operation should account for bound variables. Consider the tuples estimate to be what this single operation would produce if only that node with its sub tree existed.

I should point out that to some extent the cardinality calculator also does these things. I don't believe it looks at limit clauses, but it does look at things like "is this variable already bound higher up in the tree" and "is this variable really a constant or is it likely to change?". It uses all sorts of tricks like taking the power of a number or dividing by some factor to tweak the numbers.

Getting the naming right wasn't really important earlier because this was always purely an internal feature (and yes, _originally_ it did produce purely a cardinality as in "number of solutions" but then SPARQL got more complicated and it evolved). If we're going to expose it publicly, we need to be specific about what the numbers actually mean. The numbers as currently produced by our cardinality calculators can _not_ be reliably interpreted as meaning "number of solutions/rows/tuples". From my point of view, the only thing you can really say about them is that they are a (crude) best guess at how expensive a particular operation is going to be. Hence 'cost estimate'.

Reply:
(Scratch what I said earlier, I had where things happened mixed in my head a bit - I deleted a previous reply because it would just have added to the confusion)

...if the number you intend to produce here is _purely_ the outcome of `EvaluationStatics.getCardinality` then yes, using something like `result size estimate` or `estimated number of solutions` is fine. In fact, if we want to avoid further naming confusion, we even just go with cardinality. `getEstimatedCardinality`?

My impression (and perhaps I was wrong there) was that what you wanted to produce was _not_ just the outcome of that estimator, but the outcome of the internal cost calculation that the QueryJoinOptimizer does based (partially) on that number.

Reply:
`Cardinality` has a fairly strict definition in mathematics of being the number of distinct elements in a set. We need a term to represent the number of elements in a **bag** because we want to be able to use the same term on other nodes in the plan. `Size` is already commonly used in java, so we could go with that. I've had a look at other SQL databases though and `rows` is very common. 

For other graph databases:
 - Neo4J: rows
 - Amazone Neptune: patternEstimate
 - Stardog: cardinality
 - Ontotext: size or iterations
 - Oracle: 10 minutes of googling and still couldn't figure it out
 - Allegrograph: cursor-estimate
 - Blazegraph: unitsOut
 - MarkLogic: maybe "result estimate"
 - Jena: couldn't find any examples

At this point we could pick "potato" and we wouldn't be any worse off. Was really hoping for a clear winner when I started down this path of googling.

Reply:
You say potato...

Id prefer sticking with cardinality because that's what is already used in the EvaluationStatistics, but I'm also fine with size (or more specifically sizeEstimate or perhaps resultSizeEstimate). Hard no on rows for me I'm afraid, really don't want to get trapped in false equivalences with relational model.

Reply:
resultSizeEstimate it is :)